lib/compiler.js

```diff
@@ -538,3 +548,33 @@ function resolveRelativePaths(relativePaths, appRootDir) {
     }
   });
 }
+
+function isPreferredExtension (filename, includeExtensions, excludeExtensions) {
+  includeExtensions = includeExtensions || Object.keys(require.extensions);
+  excludeExtensions = excludeExtensions || [FILE_EXTENSION_JSON];
+  var ext = path.extname(filename);
+  return ((includeExtensions.indexOf(ext) > -1) &&
+    (excludeExtensions.indexOf(ext) === -1));
```

Comment:
You are converting a fast hashmap access O(1) to a slow array scan O(n). Also considering that the two arguments are not used, I find it suboptimal to convert an already-built lookup table in `require.extensions` to an array.

```js
return (ext in includeExtensions) && !(ext in excludeExtensions);
```